consensus/transaction_solidifier/transaction_solidifier.c

```diff
@@ -0,0 +1,383 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
+ * https://github.com/iotaledger/entangled
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
+#include "transaction_solidifier.h"
+
+#include "consensus/utils/tangle_traversals.h"
+#include "utils/hash_containers.h"
+#include "utils/logger_helper.h"
+
+#define TRANSACTION_SOLIDIFIER_LOGGER_ID "consensus_transaction_solidifier"
+
+#define SOLID_PROPAGATION_INTERVAL 500000
+
+/*
+ * Forward declarations
+ */
+
+static flex_trit_t genesis_hash[FLEX_TRIT_SIZE_243];
+
+static retcode_t check_approvee_solid_state(transaction_solidifier_t *const ts,
+                                            flex_trit_t *const approvee,
+                                            bool *solid);
+
+static retcode_t check_transaction_and_update_solid_state(
+    transaction_solidifier_t *const ts, flex_trit_t *const transaction,
+    bool *const is_new_solid);
+
+static void *spawn_solid_transactions_propagation(void *arg);
+
+static retcode_t add_new_solid_transaction(transaction_solidifier_t *const ts,
+                                           flex_trit_t *const hash);
+
+static retcode_t propagate_solid_transactions(
+    transaction_solidifier_t *const ts);
+
+static retcode_t check_solidity_do_func(tangle_t *const tangle,
+                                        flex_trit_t *hash,
+                                        iota_stor_pack_t *pack, void *data,
+                                        bool *should_branch, bool *should_stop);
+
+/*
+ * Auxiliary structs
+ */
+
+typedef struct check_solidity_do_func_params_s {
+  transaction_solidifier_t *ts;
+  bool is_milestone;
+  bool is_solid;
+} check_solidity_do_func_params_t;
+
+static void *spawn_solid_transactions_propagation(void *arg) {
+  transaction_solidifier_t *ts = (transaction_solidifier_t *)arg;
+  while (ts->running) {
+    if (propagate_solid_transactions(ts) != RC_OK) {
+      log_critical(TRANSACTION_SOLIDIFIER_LOGGER_ID,
+                   "Failed in solid transaction propagation %s\n");
+    }
+    usleep(SOLID_PROPAGATION_INTERVAL);
+  }
+  return NULL;
+}
+
+static retcode_t propagate_solid_transactions(
+    transaction_solidifier_t *const ts) {
+  retcode_t ret;
+  hash243_set_t transactions_to_propagate;
+  lock_handle_lock(&ts->lock);
+  if ((ret = hash243_set_for_each(&ts->newly_set_solid_transactions,
```

Comment:
You can also use `hash243_set_append` on a `NULL` set

Reply:
That make me think that we can re-implement `hash243_set_append` with `hash243_set_for_each `